postgres-persistence/src/main/java/com/netflix/conductor/postgres/util/PostgresQueueListener.java

```diff
@@ -0,0 +1,207 @@
+
+    protected final Logger logger = LoggerFactory.getLogger(getClass());
+
+    public PostgresQueueListener(DataSource dataSource, PostgresProperties properties) {
+        logger.info("Using experimental PostgresQueueListener");
+        this.dataSource = dataSource;
+        this.stalePeriod = properties.getExperimentalQueueNotifyStalePeriod();
+        connect();
+    }
+
+    public boolean hasMessagesReady(String queueName) {
+        checkUpToDate();
+        handleNotifications();
+        if (notificationIsStale() || !connected) {
+            connect();
+            return true;
+        }
+
+        QueueStats queueStats = queues.get(queueName);
+        if (queueStats == null) {
+            return false;
+        }
+
+        if (queueStats.getNextDelivery() > System.currentTimeMillis()) {
+            return false;
+        }
+
+        return true;
+    }
+
+    public Integer getSize(String queueName) {
+        checkUpToDate();
+        handleNotifications();
+        if (notificationIsStale() || !connected) {
+            connect();
+            return null;
+        }
+
+        QueueStats queueStats = queues.get(queueName);
+        if (queueStats == null) {
+            return 0;
+        }
+
+        return queueStats.getDepth();
+    }
+
+    private boolean notificationIsStale() {
+        return System.currentTimeMillis() - lastNotificationTime > this.stalePeriod;
+    }
+
+    private void connect() {
+        if (!connecting) {
+            connected = false;
+            connecting = true;
+            try {
+                this.conn = dataSource.getConnection();
+                this.pgconn = conn.unwrap(PGConnection.class);
+
+                try {
+                    boolean previousAutoCommitMode = conn.getAutoCommit();
+                    conn.setAutoCommit(true);
+                    try {
+                        conn.prepareStatement("LISTEN conductor_queue_state").execute();
+                    } catch (Throwable th) {
+                        conn.rollback();
+                        logger.info(th.getMessage());
+                    } finally {
+                        conn.setAutoCommit(previousAutoCommitMode);
+                        connected = true;
+                    }
+                } catch (SQLException ex) {
+                    throw new NonTransientException(ex.getMessage(), ex);
+                }
+                requestStats();
+
+            } catch (SQLException e) {
+                logger.info(e.getSQLState());
+            }
+            connecting = false;
+        }
+    }
+
+    private void requestStats() {
+        try {
+            boolean previousAutoCommitMode = conn.getAutoCommit();
+            conn.setAutoCommit(true);
+            try {
+                conn.prepareStatement("SELECT queue_notify()").execute();
+                connected = true;
+            } catch (Throwable th) {
+                conn.rollback();
+                logger.info(th.getMessage());
```

Comment:
Leftover debug message?

Reply:
No - should be logger.error instead though